source/time_stepping/lapusta_time_step.cc

```diff
@@ -0,0 +1,183 @@
+            in.reinit(fe_values,
+                      cell,
+                      this->introspection(),
+                      this->get_solution());
+
+            this->get_material_model().evaluate(in, out);
+
+            fe_values[this->introspection().extractors.velocities].get_function_values (this->get_solution(),
+                                                                                        velocity_values);
+            double max_local_velocity = 0;
+            const double delta_x = cell->minimum_vertex_distance();
+            for (unsigned int q=0; q<n_q_points; ++q)
+              {
+                // TODO in Lapusta, this should be plastic velocity. But just taking the full velocity
+                // should be the most conservative approach, so it should be ok...
+                const double local_velocity = velocity_values[q].norm();
+
+                std::pair<double,double> delta_theta_max_and_critical_slip_distance = viscoplastic.compute_delta_theta_max(
+                                                                                        in.composition[0],  // should this be 0 or q? for all the times I hand over composition.. found 0 in some part of the code, but q intuitively makes more sense
+                                                                                        in.position[q],
+                                                                                        delta_x,
+                                                                                        in.pressure[q]);
+
+                min_state_weakening_time_step = std::min (min_state_weakening_time_step,
+                                                          delta_theta_max_and_critical_slip_distance.first
+                                                          * delta_theta_max_and_critical_slip_distance.second / local_velocity);
+
+                // state healing time step is: Deltat_h = 0.2 * theta.
+                min_healing_time_step = std::min (min_healing_time_step,
+                                                  viscoplastic.compute_min_healing_time_step(in.composition[q]));
+
+                // the maximum local velocity needed for the displacement time step
+                max_local_velocity = std::max (max_local_velocity,
+                                               local_velocity);
+
+                // the viscoelastoplastic relaxation time step using the relaxation time scale: f_max * viscoplastic viscosity / shear modulus
+                // with f_max = 0.2 in Herrendörfer et al. 2018
+                // to capture the increasing slip rate in case of a purely rate-dependent friction, i.e. if b = 0
+                min_vep_relaxation_time_step = std::min (min_vep_relaxation_time_step,
+                                                         0.2 * out.viscosities[q]
+                                                         / viscoplastic.get_elastic_shear_modulus(in.composition[0]));
```

Comment:
thanks for spotting this! I think the 0 is there because I tried to find out in other plugins how to access values and somewhere I saw and copied the 0